File: go/tracing.go
package otelhelper

import (
	"context"
	"crypto/tls"
	"fmt"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
	"google.golang.org/grpc/credentials"
)

func configureTracing(ctx context.Context, res *resource.Resource, opts *Options) (*sdktrace.TracerProvider, error) {
	tpOpts := []sdktrace.TracerProviderOption{
		sdktrace.WithResource(res),
	}

	if os.Getenv(EnvTracesSampler) != "" && opts.SampleRatio >= 1.0 {
		// No explicit ratio in code: standard SDK env config wins. NewTracerProvider
		// without WithSampler reads OTEL_TRACES_SAMPLER / OTEL_TRACES_SAMPLER_ARG itself.
	} else {
		var rootSampler sdktrace.Sampler
		if opts.SampleRatio >= 1.0 {
			rootSampler = sdktrace.AlwaysSample()
		} else {
			rootSampler = sdktrace.TraceIDRatioBased(opts.SampleRatio)
		}
		tpOpts = append(tpOpts, sdktrace.WithSampler(sdktrace.ParentBased(rootSampler)))
	}

	if opts.OtelEndpoint != "" {
		// OTLP push — export spans to collector, over gRPC or HTTP/protobuf
		// depending on the resolved protocol (design: config.go's
		// resolvedOtlpProtocol — explicit > OTEL_EXPORTER_OTLP_PROTOCOL >
		// port 4318 inference > grpc default).
		var exporter *otlptrace.Exporter
		var err error
		if opts.resolvedOtlpProtocol() == ProtocolHTTP {
			httpOpts := []otlptracehttp.Option{
				otlptracehttp.WithEndpoint(opts.OtelEndpoint),
				otlptracehttp.WithTimeout(time.Duration(opts.ExportTimeoutMs) * time.Millisecond),
				otlptracehttp.WithCompression(otlptracehttp.GzipCompression),
			}
			if opts.Insecure {
				httpOpts = append(httpOpts, otlptracehttp.WithInsecure())
			} else {
				httpOpts = append(httpOpts, otlptracehttp.WithTLSClientConfig(&tls.Config{}))
			}
			exporter, err = otlptracehttp.New(ctx, httpOpts...)
		} else {
			grpcOpts := []otlptracegrpc.Option{
				otlptracegrpc.WithEndpoint(opts.OtelEndpoint),
				otlptracegrpc.WithTimeout(time.Duration(opts.ExportTimeoutMs) * time.Millisecond),
				otlptracegrpc.WithCompressor("gzip"),
			}
			if opts.Insecure {
				grpcOpts = append(grpcOpts, otlptracegrpc.WithInsecure())
			} else {
				grpcOpts = append(grpcOpts, otlptracegrpc.WithTLSCredentials(credentials.NewTLS(&tls.Config{})))
			}
			exporter, err = otlptracegrpc.New(ctx, grpcOpts...)
		}
		if err != nil {
			return nil, fmt.Errorf("trace exporter: %w", err)
		}
		tpOpts = append(tpOpts, sdktrace.WithBatcher(exporter))
	}
	// When OtelEndpoint is empty, no exporter is added. Spans still work
	// in-process for context propagation; they are simply not exported.

	if opts.DebugLevel {
		tpOpts = append(tpOpts, sdktrace.WithSpanProcessor(&debugProcessor{}))
	}

	tp := sdktrace.NewTracerProvider(tpOpts...)
	otel.SetTracerProvider(tp)
	return tp, nil
}

// StartRootSpan starts a new span detached from any parent (new trace).
// Use in workers where each iteration should be an independent trace.
func StartRootSpan(ctx context.Context, tracer trace.Tracer, name string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
	opts = append(opts, trace.WithNewRoot())
	return tracer.Start(ctx, name, opts...)
}

File: python/otel_helper/tracing.py
"""Tracing setup — equivalent to TracerSetup.cs."""

import os
from contextlib import contextmanager

from opentelemetry import context, trace
from opentelemetry.sdk.trace import TracerProvider, SpanProcessor
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.trace.sampling import ALWAYS_ON, TraceIdRatioBased
from opentelemetry.sdk.resources import Resource

from otel_helper.config import ENV_TRACES_SAMPLER, PROTOCOL_HTTP, TelemetryOptions
from otel_helper.processors import DebugProcessor

HEALTH_PATHS = frozenset(["/ping", "/health", "/healthz", "/ready"])


def configure_tracing(resource: Resource, options: TelemetryOptions) -> TracerProvider:
    """Configure and set the global TracerProvider.

    When otel_endpoint is empty, TracerProvider is configured without an OTLP exporter.
    In-process context propagation and spans still work (useful for metric exemplars).
    """
    if os.getenv(ENV_TRACES_SAMPLER) and options.sample_ratio >= 1.0:
        # No explicit ratio in code: standard SDK env config wins. TracerProvider
        # without a sampler reads OTEL_TRACES_SAMPLER / OTEL_TRACES_SAMPLER_ARG itself.
        provider = TracerProvider(resource=resource)
    else:
        sampler = ALWAYS_ON if options.sample_ratio >= 1.0 else TraceIdRatioBased(options.sample_ratio)
        provider = TracerProvider(
            resource=resource,
            sampler=sampler,
        )

    if options.otel_endpoint:
        if options.resolved_otlp_protocol() == PROTOCOL_HTTP:
            from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter as HttpSpanExporter

            # The HTTP exporter has no separate "insecure" flag — transport
            # security is purely the endpoint's own scheme (http:// vs
            # https://), unlike gRPC's independent insecure bool. An explicit
            # endpoint is used as-is by the SDK (no auto path), so the
            # standard /v1/traces suffix is appended here.
            endpoint = options.otel_endpoint.rstrip("/")
            if not endpoint.endswith("/v1/traces"):
                endpoint = f"{endpoint}/v1/traces"
            exporter = HttpSpanExporter(
                endpoint=endpoint,
                timeout=options.export_timeout_ms / 1000,
            )
        else:
            from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter

            exporter = OTLPSpanExporter(
                endpoint=options.otel_endpoint,
                insecure=options.resolve_insecure(),
                timeout=options.export_timeout_ms / 1000,
            )
        provider.add_span_processor(BatchSpanProcessor(exporter))

    if options.debug_level:
        provider.add_span_processor(DebugProcessor())

    trace.set_tracer_provider(provider)
    return provider


@contextmanager
def start_root_span(tracer, name, **kwargs):
    """Start a new root span (new trace), detaching from any current parent context.

    Use in workers/background services where each iteration should be an independent trace.
    Equivalent to .NET's ActivitySource.StartRootActivity().

    Usage:
        with start_root_span(tracer, "process.cycle") as span:
            span.set_attribute("batch.size", 10)
            ...
    """
    from opentelemetry.trace import NonRecordingSpan, INVALID_SPAN_CONTEXT
    clean_ctx = trace.set_span_in_context(NonRecordingSpan(INVALID_SPAN_CONTEXT))
    token = context.attach(clean_ctx)
    try:
        with tracer.start_as_current_span(name, **kwargs) as span:
            yield span
    finally:
        context.detach(token)
